apps/builder/app/builder/features/props-panel/use-props-logic.ts

```diff
@@ -1,156 +1,240 @@
 import { useState } from "react";
 import { nanoid } from "nanoid";
 import type { Instance, Prop } from "@webstudio-is/project-build";
-import type { WsComponentPropsMeta } from "@webstudio-is/react-sdk";
 import { getComponentPropsMeta } from "@webstudio-is/react-sdk";
+import type { PropMeta, PropValue } from "./shared";
+
+type PropOrName = { prop?: Prop; propName: string };
+export type PropAndMeta = { prop?: Prop; propName: string; meta: PropMeta };
+export type NameAndLabel = { name: string; label?: string };
+
+// The value we set prop to when it's added
+export const getStartingValue = (meta: PropMeta): PropValue | undefined => {
+  if (meta.type === "string" && meta.control !== "file-image") {
+    return {
+      type: "string",
+      value: meta.defaultValue ?? "",
+    };
+  }
 
-export type UserPropValue = Prop extends infer T
-  ? T extends { value: unknown; type: unknown }
-    ? { value: T["value"]; type: T["type"] }
-    : never
-  : never;
-
-export const getValueFromPropMeta = (
-  propValue?: WsComponentPropsMeta["props"][string]
-) => {
-  let typedValue: UserPropValue = {
-    type: "string",
-    value: `${propValue?.defaultValue ?? ""}`,
-  };
+  if (meta.type === "number") {
+    return {
+      type: "number",
+      value: meta.defaultValue ?? 0,
+    };
+  }
 
-  if (propValue?.control === "boolean") {
-    typedValue = {
+  if (meta.type === "boolean") {
+    return {
       type: "boolean",
-      value: propValue.defaultValue ?? false,
+      value: meta.defaultValue ?? false,
     };
   }
 
-  if (propValue?.control === "number") {
-    typedValue = {
-      type: "number",
-      value: propValue.defaultValue ?? 0,
+  if (meta.type === "string[]") {
+    return {
+      type: "string[]",
+      value: meta.defaultValue ?? [],
     };
   }
 
-  return typedValue;
+  // If undefined is returned,
+  // we will not add a prop in storage until we get an onChange from control.
+  //
+  // User may have this experience:
+  //   - they added a prop but didn't touch the control
+  //   - they closed props panel
+  //   - when they open props panel again, the prop is not there
+  //
+  // We want to avoid this if possible, but for some types like "asset" we can't
+  return undefined;
```

Comment:
Why would your return an undefined, its the default behavior? Needed someplace for the comment? How about about the function declaration

Reply:
Ah. I'm just not used to use `udefined` and always forget you don't have to return it explicitly.

But yeah, in this case it's also better place for the comment IMO. Can put on the top though, not a strong opinion.